src/xdplwf/recv.c

```diff
@@ -565,64 +642,693 @@ XdpGenericReceiveEnqueueTxNb(
+                (NDIS_TCP_IP_CHECKSUM_NET_BUFFER_LIST_INFO *)
+                    &NET_BUFFER_LIST_INFO(TxNbl, TcpIpChecksumNetBufferListInfo);
+            UINT16 TcpHeaderAndPayloadLength;
+
+            SegmentSize = min(TcpUserDataLen, LsoMss);
+            MssInSegment = 1;
+            TcpHeaderAndPayloadLength = (UINT16)(TcpTotalHdrLen - TcpOffset + SegmentSize);
+
+            ASSERT(Csum->Value == NULL);
+            Csum->Transmit.TcpChecksum = TRUE;
+            Csum->Transmit.TcpHeaderOffset = TcpOffset;
+            TxTcp->th_sum =
+                XdpChecksumFold(TxTcp->th_sum + htons(TcpHeaderAndPayloadLength));
+
+            if (LsoIpVersion == NDIS_TCP_LARGE_SEND_OFFLOAD_IPv4) {
+                IPV4_HEADER *Ipv4 =
+                    RTL_PTR_ADD(NblContext->Gso.Headers, sizeof(ETHERNET_HEADER));
+
+                Ipv4->TotalLength =
+                    htons((UINT16)(TcpTotalHdrLen - sizeof(ETHERNET_HEADER) + SegmentSize));
+                Csum->Transmit.IpHeaderChecksum = TRUE;
+                Csum->Transmit.IsIPv4 = TRUE;
+            } else {
+                IPV6_HEADER *Ipv6 =
+                    RTL_PTR_ADD(NblContext->Gso.Headers, sizeof(ETHERNET_HEADER));
+
+                ASSERT(LsoIpVersion == NDIS_TCP_LARGE_SEND_OFFLOAD_IPv6);
+
+                Ipv6->PayloadLength = htons(TcpHeaderAndPayloadLength);
+                Csum->Transmit.IsIPv6 = TRUE;
+            }
+
+            if (TaskOffload == NULL || !TaskOffload->Checksum.Enabled) {
+                NeedChecksum = TRUE;
+            }
+        }
+
+        ASSERT(SegmentSize > 0 || (TcpUserDataLen == 0 && SegmentsCreated == 0));
+        ASSERT(MssInSegment > 0);
+
+        //
+        // Finalize IP and TCP header values.
+        //
+
+        TxTcp->th_seq = htonl(SeqNum);
+
+        if (SegmentSize == TcpUserDataLen) {
+            //
+            // This is the final segment.
+            //
+            TxTcp->th_flags |= (RscThFlags & TH_PSH);
+        }
+
+        if (LsoIpVersion == NDIS_TCP_LARGE_SEND_OFFLOAD_IPv4) {
+            IPV4_HEADER *Ipv4 = RTL_PTR_ADD(NblContext->Gso.Headers, sizeof(ETHERNET_HEADER));
+
+            //
+            // Update the IP identification field. The LSO spec requires
+            // segments be constrained to a subset of the range (0x0000 to
+            // 0x7FFF) but this is incompatible with arbitrary packet
+            // sources (e.g. Linux) and relies on a deprecated feature
+            // (Chimney). Violate the spec and use all 16 bits, which should
+            // be harmless.
+            //
+            // N.B. the RSC packet could have had been created from a series
+            // of packets with arbitrary IDs, including all zero IDs. Re-
+            // segmenting the packet will produce different IDs, but this
+            // should be harmless according to RFC 6864.
+            //
+            Ipv4->Identification = htons((UINT16)(ntohs(Ipv4->Identification) + SegmentsCreated));
```

Comment:
Would it be a problem if the IP ID is conflicting with the original sender when, for example, the NIC coalesced 2 packets into a RSC packet and XDP built 3 packets out of the RSC packet?

Reply:
Yeah, this could potentially be troublesome. I thought of adding more heuristics here (e.g., don't stamp the ID if it is already zero and the DF bit is set) but couldn't justify such a hand-wavy idea. I also don't know how we can solve the problem with imperfect information and the additional (but unlikely to be hit, in practice) MTU constraints.

Realistically, as long as the DF bit is set, which it typically should be, the value within the IP ID should be irrelevant. That's a lot of _shoulds_, and not musts, though.

Reply:
One idea I had is we only build X packets (including packets segmented from LSO) if the packet was coalesced from X packets. Not sure if it's a good idea or necessary so up to you.

Reply:
Yes, we already impose that constraint, which is essential, else we could produce segments larger than the original, pre-coalescing segments. That's being implicitly tested, too, but I will add an explicit `TEST_TRUE(TotalSegmentCount <= Params->GroSegCount)` to our functional test to avoid any possible arithmetic errors.